File: 2tdsr/enterprise-application-development/01-revisao/src/br/com/fiap/exercicio/bean/ContaPoupanca.java
package br.com.fiap.exercicio.bean;

import java.util.Calendar;

import br.com.fiap.exercicio.exception.SaldoInsuficienteException;

public class ContaPoupanca extends Conta implements ContaInvestimento {

	public Float taxa;
	public static final Float RENDIMENTO = 0.04f;

	@Override
	public final double calculaRetornoInvestimento() {
		return saldo * RENDIMENTO;
	}

	@Override
	public void retirar(Double valor) throws SaldoInsuficienteException {
		if (valor + taxa > saldo) {
			throw new SaldoInsuficienteException("Saldo insuficiente");
		}

		saldo -= valor + taxa;
	}

	public ContaPoupanca() {
	}

	public ContaPoupanca(Integer agencia, Integer numero, Calendar dataAbertura, Double saldo, Float taxa) {
		super(agencia, numero, dataAbertura, saldo);
		this.taxa = taxa;
	}

}

File: 2tdsr/enterprise-application-development-asp-net/Fiap.Aula01/Fiap.Aula01.Exercicio01/Models/ContaPoupanca.cs
using Fiap.Aula01.Exercicio01.Models;
using Fiap.Aula01.Exercicio01.Exceptions;

namespace Fiap.Aula01.Exercicio01.Models
{
    public class ContaPoupanca : Conta : IContaInvestimento
    {

        public decimal Taxa { get; set; }

        public ContaPoupanca(taxa decimal, int agencia, int numero, 
            IList<Cliente> clientes) : base(agencia, numero, clientes)
        {
            Taxa = taxa;
        }
        
        public override void Retirar(decimal valor)
        {
            if(valor + Taxa > Saldo)
            {
                throw new SaldoInsuficienteException("Saldo insuficiente");
            }
            Saldo -= valor;
        }

        public decimal CalcularRetornoInvestimento() 
        {
            return Saldo * 4 / 100;
        }
        
    }
}